### PR TITLE
[C++] [Qt5] Set default base path and change constructor to const ref

### PR DESCRIPTION
**Pull request by etherealjoy**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

Changes
- Change constructor signature to const reference for `QString`
- Add initializer for `basepath`, `host`
- moved common code to super

`cc:`
@ravinikam @stkrwork @MartinDelille

### CODE REVIEW

File: samples/client/petstore/cpp-qt5/client/OAIPetApi.cpp [hunk 1]
```diff
@@ -91,7 +91,7 @@ OAIPetApi::deletePet(const qint64& pet_id, const QString& api_key) {
     OAIHttpRequestInput input(fullPath, "DELETE");
 
     if (api_key != nullptr) {
-        input.headers.insert("api_key", "api_key");
+        input.headers.insert("api_key", api_key);
```
Comment (stkrwork) [inline samples/client/petstore/cpp-qt5/client/OAIPetApi.cpp]: this change is not part of this PR. Were the samples not regenerated in the PR that performed this change?

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache [hunk 2]
```diff
@@ -18,7 +18,7 @@ class {{classname}}: public QObject {
 
 public:
     {{classname}}();
-    {{classname}}(QString host, QString basePath);
+    {{classname}}(const QString& host, const QString& basePath);
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache]:
```suggestion
    {{classname}}(const QString& host, const QString& basePath = "{{{basePathWithoutHost}}}");
```

File: samples/client/petstore/cpp-qt5/client/OAIPetApi.cpp (hunk 1 above)
Comment (etherealjoy) [reply samples/client/petstore/cpp-qt5/client/OAIPetApi.cpp]: No it was not updated. But this is correct.

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache (hunk 2 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache]: Yes good idea.

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache [hunk 3]
```diff
@@ -9,15 +9,15 @@
 namespace {{this}} {
 {{/cppNamespaceDeclarations}}
 
-{{classname}}::{{classname}}() {
+{{classname}}::{{classname}}() : basePath("{{{basePathWithoutHost}}}"){
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: Whats the default value of host?

Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: Yes, I need to add this.
I thought of making these members private but it might break some things.

Comment (stkrwork) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: Should the user really be able to change the base path and host after initializing the api?

Comment (stkrwork) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: So private members would be better i think

Comment (stkrwork) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-body.mustache]: ok. then lets add a setter

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache (hunk 2 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/api-header.mustache]: I had to revert this when I provide default values, because then it looks like the default constructor with no parameters.

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/main-api-server.mustache [hunk 4]
```diff
@@ -53,7 +53,7 @@ int main() {
     std::vector<int> sigs{SIGQUIT, SIGINT, SIGTERM, SIGHUP};
     setUpUnixSignals(sigs);
 #endif
-    Pistache::Address addr(Pistache::Ipv4::any(), Pistache::Port({{serverPort}}));
+    Pistache::Address addr(Pistache::Ipv4::any(), Pistache::Port({{#serverPort}}{{serverPort}}{{/serverPort}}{{^serverPort}}8080{{/serverPort}}));
```
Comment (stkrwork) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/main-api-server.mustache]: if you change stuff for pistache as well here, then you have to regenerate the samples for pistache as well

Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/main-api-server.mustache]: I did, there is no change. 
The method produces default 8080 `serverPort` value, now it does not so that we don't need to append port for clients.
Check the removal of code from the pistache java file.

Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/main-api-server.mustache]: I also regenerated Qt5 server, which was not updated before.

File: samples/server/petstore/cpp-pistache/.openapi-generator/VERSION [hunk 5]
```diff
@@ -1 +1 @@
-4.0.0-SNAPSHOT
\ No newline at end of file
+4.0.1-SNAPSHOT
```
Comment (etherealjoy) [inline samples/server/petstore/cpp-pistache/.openapi-generator/VERSION]: for pistahce only this changed

**Comment by wing328** [comment]
@etherealjoy thanks again for your contribution, which has been included in the v4.0.1 release (https://twitter.com/oas_generator/status/1135534738658062336)
